datafusion/core/src/physical_optimizer/replace_repartition_execs.rs

```diff
@@ -0,0 +1,778 @@
+// under the License.
+
+//! Repartition optimizer that replaces `SortExec`s and their suitable `RepartitionExec` children with `SortPreservingRepartitionExec`s.
+use datafusion_common::config::ConfigOptions;
+use datafusion_common::tree_node::{Transformed, TreeNode};
+
+use crate::error::Result;
+
+use crate::physical_plan::repartition::RepartitionExec;
+use crate::physical_plan::repartition::SortPreservingRepartitionExec;
+use crate::physical_plan::sorts::sort::SortExec;
+use crate::physical_plan::ExecutionPlan;
+
+use super::utils::is_repartition;
+use super::PhysicalOptimizerRule;
+
+use datafusion_physical_expr::utils::ordering_satisfy;
+use itertools::enumerate;
+
+use std::sync::Arc;
+
+#[derive(Default)]
+/// ReplaceRepartitionExecs optimizer rule searches for `SortExec`s and their `RepartitionExec`
+/// children with multiple input partitioning (with proper ordering) so that it can replace the `RepartitionExec`s with
+/// `SortPreservingRepartitionExec`s and remove the additional `SortExec` from the physical plan.
+/// The rule only works for conditions that have a `SortExec` and at least one `RepartitionExec`
+/// with multiple input partitioning.
+///
+/// The algorithm flow is simply like this:
+/// 1. Visit nodes of the physical plan top-down and look for a SortExec node (if not found do nothing)
+/// 2. If `SortExec` is found, iterate over its children recursively until the leaf node
+///      or check if anything breaks the ordering between,
+///      `RepartitionExec`s with multiple input partitions considered as they maintain input ordering
+///       because they're potentially changed with `SortPreservingRepartitionExec`s
+/// 2_1. If the ordering is not being maintained from input, break the cycle of node visiting
+/// 3_1. Check if the child is a `RepartitionExec` with multiple input partitions
+/// 3_1_1. If conditions match, replace the `RepartitionExec` with `SortPreservingRepartitionExec`
+/// 3_1_2. If conditions do not match, keep the plan as is.
+/// 4. End the bottom-up transformation loop
+/// 5. Check if the `SortExec` plan is still necessary by comparing its inputs ordering and `SortExec` ordering
+/// 5_1. If ordering is the same, remove the `SortExec` from the plan
+/// 5_2. Otherwise keep the plan as is.
+/// 6. Continue to the top-down transformation
+pub struct ReplaceRepartitionExecs {}
+
+impl ReplaceRepartitionExecs {
+    #[allow(missing_docs)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+/// Creates a `SortPreservingRepartitionExec` from given `RepartitionExec`
+fn sort_preserving_repartition(
+    repartition: &RepartitionExec,
+) -> Result<Arc<SortPreservingRepartitionExec>> {
+    Ok(Arc::new(SortPreservingRepartitionExec::try_new(
+        repartition.input().clone(),
+        repartition.partitioning().clone(),
+    )?))
+}
+
+fn does_plan_maintains_input_order(plan: &Arc<dyn ExecutionPlan>) -> bool {
+    return plan.maintains_input_order().iter().all(|flag| *flag);
```

Comment:
```suggestion
    plan.maintains_input_order().iter().any(|flag| *flag)
```
using any is better, because if one of the child maintains ordering we want to continue analysis anyway